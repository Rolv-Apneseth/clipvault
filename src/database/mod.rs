use std::{path::Path, sync::LazyLock};

use include_dir::{Dir, include_dir};
use miette::{Context, IntoDiagnostic, Result};
use rusqlite::Connection;
use rusqlite_migration::Migrations;
use tracing::instrument;

pub mod data;
pub mod queries;

// DB MIGRATIONS, DEFINED IN ./migrations
static MIGRATIONS_DIR: Dir = include_dir!("$CARGO_MANIFEST_DIR/src/database/migrations");
static MIGRATIONS: LazyLock<Migrations<'static>> =
    LazyLock::new(|| Migrations::from_directory(&MIGRATIONS_DIR).unwrap());

/// Get a database connection. Make sure the DB is initialised somewhere first
/// before calling this function.
pub fn get_db_connection(path_db: &Path) -> Result<Connection> {
    Connection::open(path_db)
        .into_diagnostic()
        .context("failed to connect to the database")
}

#[instrument]
/// Initialise the database, creating the file and applying migrations if needed
pub fn init_db(path_db: &Path) -> Result<Connection> {
    tracing::debug!("initialising DB");
    let mut conn = get_db_connection(path_db)?;

    tracing::trace!("applying PRAGMA");
    conn.pragma_update(None, "journal_mode", "WAL")
        .into_diagnostic()
        .context("failed to apply PRAGMA: journal mode")?;

    tracing::trace!("applying migrations");
    MIGRATIONS
        .to_latest(&mut conn)
        .into_diagnostic()
        .context("failed to apply migrations")?;

    Ok(conn)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_migrations() {
        assert!(MIGRATIONS.validate().is_ok());
        insta::assert_debug_snapshot!(MIGRATIONS);
    }
}
