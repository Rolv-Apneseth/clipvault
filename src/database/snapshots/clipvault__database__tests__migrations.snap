---
source: src/database/mod.rs
expression: MIGRATIONS
---
LazyLock(
    Migrations {
        ms: [
            M {
                up: "CREATE TABLE IF NOT EXISTS clipboard\n(\n    id integer PRIMARY KEY,\n    content blob NOT NULL UNIQUE,\n    last_updated integer NOT NULL\n) STRICT ;\n",
                up_hook: None,
                down: Some(
                    "DROP TABLE IF EXISTS clipboard CASCADE;\n",
                ),
                down_hook: None,
                foreign_key_check: false,
                comment: Some(
                    "01-clipboard",
                ),
            },
            M {
                up: "CREATE INDEX IF NOT EXISTS last_updated ON clipboard (last_updated);\n",
                up_hook: None,
                down: Some(
                    "DROP INDEX IF EXISTS last_updated;\n",
                ),
                down_hook: None,
                foreign_key_check: false,
                comment: Some(
                    "02-last-updated-index",
                ),
            },
        ],
    },
)
